File: packages/sdk-dart/lib/src/resources/phone_numbers.dart
import '../models/telephony.dart';
import '../page.dart';
import '../transport.dart';

/// Phone numbers you own at your carrier, registered on a SIP trunk.
class PhoneNumbers {
  /// Binds the resource to a transport.
  const PhoneNumbers(this._transport);

  final Transport _transport;

  /// Lists phone numbers.
  Future<Page<PhoneNumber>> list(
          {int? limit, String? startingAfter, String? endingBefore}) =>
      _transport.list('/v1/phone-numbers', PhoneNumber.fromJson, query: {
        'limit': limit,
        'starting_after': startingAfter,
        'ending_before': endingBefore,
      });

  /// Creates one.
  Future<PhoneNumber> create(CreatePhoneNumber body) =>
      _transport.request('POST', '/v1/phone-numbers', PhoneNumber.fromJson,
          body: body.toJson());

  /// Fetches one.
  Future<PhoneNumber> retrieve(String id) => _transport.request('GET',
      '/v1/phone-numbers/${Uri.encodeComponent(id)}', PhoneNumber.fromJson);

  /// Changes one. Only the fields you set are changed.
  Future<PhoneNumber> update(String id, UpdatePhoneNumber body) =>
      _transport.request('PATCH',
          '/v1/phone-numbers/${Uri.encodeComponent(id)}', PhoneNumber.fromJson,
          body: body.toJson());

  /// Deletes one.
  Future<void> delete(String id) => _transport.requestVoid(
      'DELETE', '/v1/phone-numbers/${Uri.encodeComponent(id)}');
}

File: packages/sdk-ts/src/resources/phone-numbers.ts
import type { Http, RequestOptions } from '../core/http.js';
import type { Page } from '../core/page.js';
import type { CreatePhoneNumber, PhoneNumber, PaginationQuery, UpdatePhoneNumber } from '../types.js';

const enc = encodeURIComponent;

/** Phone numbers you own at your carrier, registered on a SIP trunk. */
export class PhoneNumbers {
    constructor(private readonly http: Http) {}

    list(query: PaginationQuery = {}, opts?: RequestOptions): Promise<Page<PhoneNumber>> {
        return this.http.list<PhoneNumber>('/v1/phone-numbers', query, opts);
    }

    create(body: CreatePhoneNumber, opts?: RequestOptions): Promise<PhoneNumber> {
        return this.http.request<PhoneNumber>('POST', '/v1/phone-numbers', { ...opts, body });
    }

    retrieve(id: string, opts?: RequestOptions): Promise<PhoneNumber> {
        return this.http.request<PhoneNumber>('GET', `/v1/phone-numbers/${enc(id)}`, opts);
    }

    /** Only the fields sent are changed. */
    update(id: string, body: UpdatePhoneNumber, opts?: RequestOptions): Promise<PhoneNumber> {
        return this.http.request<PhoneNumber>('PATCH', `/v1/phone-numbers/${enc(id)}`, { ...opts, body });
    }

    delete(id: string, opts?: RequestOptions): Promise<void> {
        return this.http.request<void>('DELETE', `/v1/phone-numbers/${enc(id)}`, opts);
    }
}
